test/libsolidity/util/compiler/InternalCompiler.cpp

```diff
@@ -0,0 +1,150 @@
+            LinkerObject const& runtimeObject = m_stack.runtimeObject(contractName);
+            bool hasUnlinkedReferences = !object.linkReferences.empty();
+            bytes cborMetadata = m_stack.cborMetadata(contractName);
+
+            std::optional<AssemblyItems> assemblyItems;
+            std::optional<AssemblyItems> runtimeAssemblyItems;
+
+            if (!_input.eofVersion.has_value())
+            {
+                if (auto items = m_stack.assemblyItems(contractName); items != nullptr)
+                    assemblyItems = std::make_optional(*items);
+
+                if (auto items = m_stack.runtimeAssemblyItems(contractName); items != nullptr)
+                    runtimeAssemblyItems = std::make_optional(*items);
+            }
+
+            if (!m_stack.isExperimentalSolidity())
+                compiled.emplace_back(CompiledContract{
+                    contractName,
+                    object.bytecode,
+                    runtimeObject.bytecode,
+                    hasUnlinkedReferences,
+                    cborMetadata,
+                    assemblyItems,
+                    runtimeAssemblyItems,
+                    m_stack.metadata(contractName),
+                    m_stack.contractABI(contractName),
+                    m_stack.interfaceSymbols(contractName),
+                    generateEventSignatures(contractName)
+                });
+            else
+                compiled.emplace_back(CompiledContract{
+                    contractName,
+                    object.bytecode,
+                    runtimeObject.bytecode,
+                    hasUnlinkedReferences,
+                    cborMetadata,
+                    assemblyItems,
+                    runtimeAssemblyItems,
+                    std::nullopt,
+                    std::nullopt,
+                    std::nullopt,
+                    std::vector<EventSignature>{}
+                });
+        }
+        mappedContracts.insert(std::make_pair(sourceName, compiled));
+    }
+
+    return std::make_pair(mappedContracts, m_stack.errors());
+}
+
+std::vector<EventSignature> InternalCompiler::generateEventSignatures(
+    std::string const& _contractName
+) const
+{
+    auto const& contract = m_stack.contractDefinition(_contractName);
+    auto const& events = contract.events();
+    auto const& interfaceEvents = contract.usedInterfaceEvents();
+
+    auto toSignature = [](EventDefinition const* _event)
+    {
+        soltestAssert(_event);
+        return EventSignature::fromEvent(*_event);
+    };
+
+    auto signatures =
+        ranges::views::concat(events, interfaceEvents) |
+        ranges::views::transform(toSignature) |
+        ranges::to<std::vector>();
+
+    return signatures;
+}
+
+std::string InternalCompiler::formatErrorInformation() const
+{
+    std::string errorInformation;
+    for (auto const& error: m_stack.errors())
+    {
+        auto formatted = SourceReferenceFormatter::formatErrorInformation(
```

Comment:
I'd cram that directly into the `errorInformation.append`. Then you're not prone of making another copy here. It's almost certainly being optimized away but yeah, not entirely certain either :)